graphql_compiler/query_planning/typedefs.py

```diff
@@ -0,0 +1,310 @@
+# Copyright 2021-present Kensho Technologies, LLC.
+from abc import ABCMeta
+from dataclasses import dataclass, field
+from enum import Enum, unique
+from typing import Any, Callable, Dict, Iterable, Mapping, Optional, cast
+
+from dataclasses_json import DataClassJsonMixin, config
+from graphql import (
+    GraphQLList,
+    GraphQLNonNull,
+    GraphQLNullableType,
+    GraphQLScalarType,
+    GraphQLType,
+    ListTypeNode,
+    NamedTypeNode,
+    NonNullTypeNode,
+    TypeNode,
+    parse_type,
+    specified_scalar_types,
+)
+
+from .. import GraphQLDate, GraphQLDateTime, GraphQLDecimal
+from ..compiler.compiler_frontend import OutputMetadata
+from ..global_utils import is_same_type
+from ..typedefs import QueryArgumentGraphQLType
+
+
+QueryExecutionFunc = Callable[
+    [str, Dict[str, QueryArgumentGraphQLType], Dict[str, Any]], Iterable[Mapping[str, Any]]
+]
+
+# Custom scalar types.
+CUSTOM_SCALAR_TYPES = {
+    GraphQLDate.name: GraphQLDate,
+    GraphQLDateTime.name: GraphQLDateTime,
+    GraphQLDecimal.name: GraphQLDecimal,
+}
+
+# Custom scalar types must not have name conflicts with builtin scalar types.
+if set(CUSTOM_SCALAR_TYPES).intersection(specified_scalar_types):
+    raise AssertionError(
+        f"Custom scalar types must have different names than builtin scalar types. Received "
+        f"overlapping type(s) {set(CUSTOM_SCALAR_TYPES).intersection(specified_scalar_types)}. "
+        f"Custom scalar types: {set(CUSTOM_SCALAR_TYPES)}. Builtin scalar types: "
+        f"{set(specified_scalar_types)}."
+    )
+
+# Custom scalar types combined with builtin scalar types represent all allowable scalar types.
+ALL_SCALAR_TYPES = CUSTOM_SCALAR_TYPES.copy()
+ALL_SCALAR_TYPES.update(specified_scalar_types)
+
+
+def _get_type_from_scalar_type_dictionary(
+    scalar_types: Dict[str, GraphQLScalarType], type_node: TypeNode
+) -> GraphQLType:
+    """Get the GraphQL type definition from an AST node.
+
+    Given a scalar type dictionary and an AST node describing a type, return a GraphQLType
+    definition, which applies to that type. For example, if provided the parsed AST node for
+    `[Date]`, a GraphQLList instance will be returned, containing the type called
+    "Date" found in the scalar type dictionary. If a type called "Date" is not found in the scalar
+    type dictionary, then None will be returned.
+
+    Note: this is very similar to GraphQL's type_from_ast. However, instead of requiring a GraphQL
+    schema this function requires a dictionary of the scalar types. This simplifies deserialization
+    and allows for custom scalar types without constructing an entire schema.
```

Comment:
copying over this comment on #1008 from @obi1kenobi 
> Are we going to have issues with this approach if a provider wants to add some custom scalar types that the compiler doesn't natively support?
> 
> Not necessarily a deal-breaker, just want to make sure we consider the costs and benefits and explicitly choose the tradeoffs that are best here.

I think that yes, we will have an issue for custom scalar types not in the compiler. However, I'm not sure what the alternative is since we don't want deserialization to be dependent on the schema. Open to suggestions!

Reply:
Let's just document it for now as a limitation of the current approach, and revisit when we need more custom scalar types?